ambry-tools/src/main/java/com.github.ambry.tools/admin/ConcurrencyTestTool.java

```diff
@@ -0,0 +1,946 @@
+ *
+ * Caution: We might need to be cautious about OOM errors as this tool holds all the blobs(that were injected) in memory
+ * injected until the GETs have reached the threshold for the a particular blob.
+ *
+ * Supported features
+ * - Parallel Puts and Gets
+ * - Concurrent GETs for the same blobs
+ * - This also means that we will get occasional bursts of GETs
+ *
+ * TODO: Metrics
+ *
+ * Future work:
+ * - Adding delay between concurrent GETs for same blobs
+ * - Adding delay for GETs to start so that the file cache will only have recently uploaded blobs(and not all)
+ *
+ * Example :
+ *  java -cp ambry.jar com.github.ambry.tools.admin.ConcurrencyTestTool --hardwareLayout [HardwareLayoutFile]
+ *  --partitionLayout [PartitionLayoutFile] --putGetHelperFactory com.github.ambry.tools.admin.ServerPutGetHelperFactory
+ *  --maxParallelPutCount 2 --parallelGetCount 4 --burstCountForGet 4 --maxGetCountPerBlob 10
+ *  --totalPutBlobCount 1000 --maxBlobSizeInBytes 1000 --minBlobSizeInBytes 100 --hostName [HostName] --port [PortNo]
+ *  --deleteOnExit true --enableVerboseLogging true
+ *
+ *  Few other options that might be useful:
+ * --outFile [FilePath]                 : Output file to redirect the output to
+ * --sleepTimeBetweenBatchGetsInMs long : Time to sleep in ms between batch Gets per Consumer thread
+ * --sleepTimeBetweenBatchPutsInMs long : Time to sleep in ms between batch Puts per Producer thread
+ * --burstThreadCountForPut int         : No of concurrent PUTs by each producer thread
+ */
+public class ConcurrencyTestTool {
+
+  private static final Logger logger = LoggerFactory.getLogger(ConcurrencyTestTool.class);
+
+  public static void main(String args[])
+      throws Exception {
+    InvocationOptions options = new InvocationOptions(args);
+    ClusterMap clusterMap = new ClusterMapManager(options.hardwareLayoutFilePath, options.partitionLayoutFilePath,
+        new ClusterMapConfig(new VerifiableProperties(new Properties())));
+
+    PutGetHelperFactory putGetHelperFactory =
+        Utils.getObj(options.putGetHelperFactoryStr, new Properties(), options.hostName, options.port, clusterMap,
+            options.maxBlobSizeInBytes, options.minBlobSizeInBytes, options.enabledVerboseLogging);
+
+    PutGetHelper putGetHelper = putGetHelperFactory.getPutGetHelper();
+    ConcurrencyTestTool concurrencyTestTool = new ConcurrencyTestTool();
+    concurrencyTestTool.startTest(putGetHelper, options.maxParallelPutCount, options.parallelGetCount,
+        options.totalPutBlobCount, options.maxGetCountPerBlob, options.burstCountForGet, options.deleteOnExit,
+        options.enabledVerboseLogging, options.sleepTimeBetweenBatchPutsInMs, options.sleepTimeBetweenBatchGetsInMs);
+  }
+
+  /**
+   * Initializes and starts the concurrent test tool.
+   * @param putGetHelper the {@link PutGetHelper} to be used for PUTs and GETs
+   * @param maxParallelPutCount maximum number of parallel puts per phase
```

Comment:
During every put phase, we try to make some number(lets say N) of parallel put requests. This N can go upto a max of maxParallelPutCount. If we don't want this non-determinism, we can make it do N parallel put requests every time.